[VSC-10] Add support for finding references

heitor-lassarote opened the pull request:
Depends on https://github.com/serokell/motoko/pull/50 and https://github.com/serokell/node-motoko/pull/8.

* Supports ID nodes in the language server.
* Implements finding references. Extra care is given for references of `f` in `o.f` where `o` is some object or module.
* Improves finding definitions for the same cases as references.

heitor-lassarote committed:
[VSC-10] Support type search for fields

Problem: To look for `f` in `o.f`, we should support checking for `f` in
the type of `o`.

Solution: Enable combining sources in the typechecker and consider
`typeRep` from the compiler and look for matching fields when searchnig
for definition in a `DotE` expression. Add a lot of tests.

heitor-lassarote committed:
[VSC-10] Support multiple definitions

Problem: We can find multiple definitions from a type search.

Solution: Change `findDefinition` to return a list of definitions.
Rename to `findDefinitions`. Propagate changes while trying to not be
very intrusive to existing code.

heitor-lassarote committed:
[VSC-10] Add a test for corner cases

Problem: We should test that with corner cases we get some reasonable
locations.

Solution: Add a test for a more complex file.

heitor-lassarote committed:
[VSC-10] Reorder node priorities

Problem: Node priorities could be ordered from greatest to lower
priority.

Solution: Move ImportE to the end.

heitor-lassarote commented on src/server/navigation.ts:
```diff
@@ -116,19 +145,22 @@ function findNameInPattern(
+    ID: 1, // name
     PathT: 2, // type reference
     ImportE: 1, // module import
     VarP: 2, // field import
```
It wasn't sorted by priority, though... :thinking: Anyway, I sorted by priority now. Also, I removed the case for `ID` since apparently that wasn't really needed. Given how `findMostSpecificNodeForPosition`, though, `ID` might still be returned.

GoPavel commented on src/server/ast.ts:
```diff
@@ -88,6 +88,7 @@ export default class AstResolver {
             let immediateImports: string[];
             try {
                 if (typed) {
+                    motoko.setTypecheckerCombineSrcs(true);
```
Maybe we can do it in the constructor of `Context`? E.g.:
```ts
export class Context {
      constructor(uri: string, version: Version, motoko: Motoko) {
        motoko.setTypecheckerCombineSrcs(true);
        this.motoko = motoko;
        ...
    }
}
```

heitor-lassarote replied:
I'm not sure this is a better place. Now, the constructor does some side effects that don't seem good. Next to `parseMotokoTyped` it seemed better because it provided some context why we were setting to true.
I put it in `requestMotokoInstance`, that looks a bit better imo, wdyt?

GoPavel replied:
Good

heitor-lassarote committed:
[VSC-10] Include more definitions in references

Problem: It's possible that when we look for definitions for the
references in the AST, we might encounter more definitions we didn't see
before.

Solution: If the context asks to include declarations, then also include
the locations for the IDs of the new definitions.

GoPavel commented on src/server/definition.spec.ts:
let's merge this file with the `src/server/test/definition.spec.ts`

heitor-lassarote replied:
Done

GoPavel commented on test/references/circular.mo:
Maybe let's unify folders `test/defintions` and `test/references`

heitor-lassarote replied:
Done

heitor-lassarote committed:
[VSC-10] Refactor test suites

Problem: We have two files for testing definitions, so we should join
into one. Moreover, we have a lot of shared things between the
definition and reference tests.

Solution: Join test cases from `src/server/definition.spec.ts` into
`src/server/test/definition.spec.ts`. Join Motoko reference test files
tests into the Motoko definitions test files. Extract common parts to
helpers. Move reference tests to `src/server/test` directory.
